core/shared/src/main/scala/fs2/retry.scala

```diff
@@ -0,0 +1,57 @@
+package fs2
+
+import cats.effect.Async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+object retry {
+  /**
+   * Retries `fa` on failure, returning a single element
+   * stream with the result of `fa` as soon as it succeeds.
+   *
+   * @param delay Duration of delay before the first retry
+   *
+   * @param nextDelay Applied to the previous delay to compute the
+   *                  next, e.g. to implement exponential backoff
+   *
+   * @param maxRetries Number of attempts before failing with the
+   *                   latest error, if `fa` never succeeds
+   *
+   * @param retriable Function to determine whether a failure is
+   *                  retriable or not, defaults to retry every
+   *                  `NonFatal`. A failed stream is immediately
+   *                  returned when a non-retriable failure is
+   *                  encountered
+   */
+  def retry[F[_], A](fa: F[A],
+                     delay: FiniteDuration,
```

Comment:
I am not so sure by this retry combinator. It has a lot of parameters and seems to me quite heavy. 
But the main objection I have is  the fact that `f` here cannot have safe resource allocation we should not rather instead of `F` use Stream[F, A]. You can always do Stream.eval() if its only `F` you need to evaluate. 

May I suggest alternative signature here, that will be on Stream[F, A]: 

```scala 

/** runs this stream until `f` returns None. Otherwise waits given timeout before retrying. Int has number of attempts since last failure. Emits failure on left when fails.  **/
def retryWhen(f: (Throwable, Int) => F[Option[FiniteDuration]] ) : Stream[F, Either[Throwable, A]] = ??? 

def retryAlways(pause: FiniteDuration): Stream[F, Either[Throwable, A]] = 
  retryWhen { _ => F.pure(Some(period)) }

```